### PR TITLE
Add support for branch- and branding-based feature flagging

### PR DESCRIPTION
**Pull request by DHowett**
This pull request implements a "feature flagging" system that will let
us turn Terminal and conhost features on/off by branch, "release" status
or branding (Dev, Preview, etc.).

It's loosely modelled after the Windows OS concept of "Velocity," but
only insofar as it is driven by an XML document and there's a tool that
emits a header file for you to include.

It only supports toggling features at compile time, and the feature flag
evaluators are intended to be fully constant expressions.

Features are added to `src\features.xml` and marked with a "stage". For
now, the only stages available are `AlwaysDisabled` and `AlwaysEnabled`.
Features can be toggled to different states using branch and branding
tokens, as documented in the included feature flag docs.

For a given feature Feature_XYZ, we will emit two fixtures visible to
the compiler:

1. A preprocessor define `TIL_FEATURE_XYZ_ENABLED` (usable from MIDL,
   C++ and C)
2. A feature class type `Feature_XYZ` with a static constexpr member
   `IsEnabled()` (usable from C++, designed for `if constexpr()`).

Like Velocity, we rely on the compiler to eliminate dead code caused by
things that compile down to `if constexpr (false)`. :)

Michael suggested that we could use `WindowsInbox` as a branding to
determine when we were being built inside Windows to supplant our use of
the `__INSIDE_WINDOWS` preprocessor token. It was brilliant.

Design Decisions
----------------

* Emitting the header as part of an MSBuild project
   * WHY: This allows the MSBuild engine to ensure that the build is
     only run once, even in a parallel build situation.
* Only having one feature flag document for the entire project
   * WHY: Ease.
* Forcibly including `TilFeatureStaging` with `/FI` for all CL compiler
  invocations.
   * WHY: If this is a project-wide feature system, we should make it as
     easy as possible to use.
* Emitting preprocessor definitions instead of constexpr/consteval
   * WHY: Removing entire functions/includes is impossible with `if
     constexpr`.
   * WHY: MIDL cannot use a `static constexpr bool`, but it can rely on
     the C preprocessor to remove text.
* Using MSBuild to emit the text instead of PowerShell
   * WHY: This allows us to leverage MSBuild's `WriteOnlyWhenDifferent`
     task parameter to avoid changing the file's modification time when
     it would have resulted in the same contents. This lets us use the
     same FeatureStaging header across multiple builds and multiple
     branches and brandings _assuming that they do not result in a
     feature flag change_.
   * The risk in using a force-include is always that it, for some
     reason, determines that the entire project is out of date. We've
     gone to great lengths to make sure that it only does so if the
     features _actually materially changed_.

### CODE REVIEW

**Comment by DHowett** [comment]
# DESIGN DECISION DISCUSSION TOPICS

* Emitting the header as part of an MSBuild project
   * **WHY**: This allows the MSBuild engine to ensure that the build is
     only run **once**, even in a parallel build situation.
* Only having one feature flag document for the entire project
   * **WHY**: Ease.
* Forcibly including `TilFeatureStaging` with `/FI` for all CL compiler
  invocations.
   * **WHY**: If this is a project-wide feature system, we should make it as
     easy as possible to use.
* Preprocessor definitions instead of constexpr/consteval
   * **WHY**: Removing entire functions is impossible with `if constexpr`.
   * **WHY**: MIDL cannot use a `static constexpr bool`, but it can rely on
     the C preprocessor to remove text.

**Comment by DHowett** [comment]
DESIGN DECISION:

Using MSBuild to emit the text, instead of PowerShell. This allows us to leverage MSBuild's `WriteOnlyWhenDifferent` (sic) task parameter to avoid changing the file's modification time when it would have resulted in the same contents. This lets us use the same FeatureStaging header across multiple builds and multiple branches and brandings _assuming that they do not result in a feature flag change_.

The risk in using a force-include is always that it, for some reason, determines that the entire project is out of date. We've gone to great lengths to make sure that it only does so if the features _actually materially changed_.

File: doc/feature_flags.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+*always* disabled in Release, even if they come from a branch that would have
+been enabled by the wildcard.
+
+### Precedence
+
+1. `alwaysDisabledReleaseTokens`
+2. Enabled branches
+3. Disabled branches
+   * The longest branch token that matches your branch will win.
+3. Enabled brandings
```
Comment (miniksa) [inline doc/feature_flags.md:63]: weird numbering

File: build/rules/GenerateFeatureFlags.proj [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    <!--
+      We gather the feature flag output in MSBuild and emit the file so that we can take advantage of
+      WriteOnlyWhenDifferent. Doing this ensures that we don't rebuild the world when the branch changes
+      (if it results in a new TilFeatureStaging.h that would have had the same content/features as the previous one)
+    -->
+
+    <WriteLinesToFile File="$(OpenConsoleCommonOutDir)\inc\TilFeatureStaging.h"
+      Lines="@(_FeatureFlagFileLines)"
+      Overwrite="true"
+      WriteOnlyWhenDifferent="true" />
```
Comment (miniksa) [inline build/rules/GenerateFeatureFlags.proj:82]: This is some straight up wizardry.

File: doc/feature_flags.md (hunk 1 above)
Comment (DHowett) [reply doc/feature_flags.md:63]: Thanks.

**Review by miniksa — COMMENTED** [review]
Looks good to me, including the conversions from `__INSIDE_WINDOWS`. Now they actually have intent. If you've rid us of them all, though, you may want to remove that def completely so someone doesn't accidentally use it going forward and migrates to the feature flagging.

**Comment by DHowett** [comment]
document: https://github.com/microsoft/terminal/blob/dev/duhowett/jerk/doc/feature_flags.md

File: src/features.xml [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0"?>
+<featureStaging xmlns="http://microsoft.com/TilFeatureStaging-Schema.xsd">
+    <!-- See doc/feature_flags.md for more info.  -->
+    <feature>
+        <name>Feature_ExampleFeat</name>
+        <description>This feature will be replaced in the next pull request.</description>
```
Comment (DHowett) [inline src/features.xml:6]: This is https://github.com/microsoft/terminal/compare/dev/duhowett/jerk...dev/duhowett/jerkify

Comment (miniksa) [reply src/features.xml:6]: LOL ok.

File: tools/FeatureStagingSchema.xsd [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                            <xs:element name="alwaysEnabledBranchTokens" type="til:branchTokenListType" minOccurs="0" />
+
+                            <xs:element name="alwaysDisabledReleaseTokens" type="til:releaseTokenListType" minOccurs="0" />
+
+                            <xs:element name="alwaysEnabledBrandingTokens" type="til:brandingTokenListType" minOccurs="0" />
+                            <xs:element name="alwaysDisabledBrandingTokens" type="til:brandingTokenListType" minOccurs="0" />
+                        </xs:all>
+                    </xs:complexType>
+
+					<xs:key name="featureBranchOverridesKey">
```
Comment (miniksa) [inline tools/FeatureStagingSchema.xsd:32]: Mixed tabs and spaces.

**Review by miniksa — APPROVED** [review]

**Review by carlos-zamora — APPROVED** [review]
This is fantastic! So excited for this to land :)

**Pull request merged**